src/learner.cc

```diff
@@ -648,6 +662,10 @@ class LearnerImpl : public Learner {
   std::vector<std::shared_ptr<DMatrix> > cache_;
 
   common::Monitor monitor_;
+
+  /*! \brief saved config keys used to restore failed worker */
+  std::set<std::string> saved_configs_ = {"max_depth", "tree_method", "dsplit",
```

Comment:
The IO is driving me crazy ...  Do you think it's possible to modify rabit serializable object from `Load(dmlc::Stream*)` to `Load(FixedLengthBuffer )`?  It's a little bit ugly to use stream with Json.  And besides, I don't think it's possible to check point a stream with unlimited size.

Reply:
as long as we make change to this interface and add new interface. I recall what they do is to write to a file and expose stream handler.
https://github.com/dmlc/dmlc-core/blob/master/include/dmlc/io.h#L132

Reply:
It's a bit more work, so what happens because of nested call to save, a stream handler were called by various of classes and append payload onto stream and flush to disk. Load is just reverse of what save does.

Reply:
I see your point, in order to decode json, we need to load all into memory first...

Reply:
following with separate pr and link to https://github.com/dmlc/xgboost/issues/3980

Reply:
https://github.com/dmlc/xgboost/issues/4852